### PR TITLE
Implement search in Timeline flame chart

### PR DESCRIPTION
**Pull request by kenzieschmoll**
- Abstract search code from memory page into mixins that can be consumed throughout DevTools - @terrylucas please take a close look at this
- Add search bar to timeline flame chart that has navigation (next, prev) and match status functionality
- Highlight matches (yellow) and "active match" (orange) in flame chart - these colors are based off of Chrome's search function
- Scroll to matches as matches and active match are updated

flame_chart_search_small

### CODE REVIEW

**Comment by devoncarew** [comment]
The UI looks great! I like the prev/next arrows, the cancel button, and the indication of where you are in the search results. Will review the code now.

File: packages/devtools_app/lib/src/timeline/timeline_controller.dart [hunk 1]
```diff
@@ -415,6 +419,23 @@ class TimelineController
...
+  void updateSearchMatches() {
+    updateMatches(matchesForSearch(search));
+  }
+
+  List<TimelineEvent> matchesForSearch(String search) {
+    if (search == null || search.isEmpty) return [];
+    final matches = <TimelineEvent>[];
+    for (final event in data.timelineEvents) {
+      breadthFirstTraversal<TimelineEvent>(event, action: (TimelineEvent e) {
+        if (e.name.toLowerCase().contains(search.toLowerCase())) {
```
Comment (devoncarew) [inline packages/devtools_app/lib/src/timeline/timeline_controller.dart:431]: as a very minor nit, you could lowercase the search term outside of this loop

**Comment by terrylucas** [comment]
I've been looking at the escape key problem.  Looks to me like Flutter has added some support where the escape key move back to a previous page (navigation).  Maybe the RawKeyboardListener shouldn't be used but something else?  Or we use another key combo SHIFT+Escape or ALT+Escape work.

File: packages/devtools_app/lib/src/charts/flame_chart.dart [hunk 2]
```diff
@@ -452,6 +487,10 @@ class ScrollingFlameChartRowState<V> extends State<ScrollingFlameChartRow>
 
   V hovered;
 
+  List<V> searchMatches = [];
+
+  V activeSearchMatch;
```
[the remark this answers is not in the document]
Comment (kenzieschmoll) [reply packages/devtools_app/lib/src/charts/flame_chart.dart]: Added to new mixin `SearchableMixin`

File: packages/devtools_app/lib/src/charts/flame_chart.dart [hunk 3]
```diff
@@ -477,6 +516,40 @@ class ScrollingFlameChartRowState<V> extends State<ScrollingFlameChartRow>
...
+            .where((match) => nodeData.contains(match))
+            .toList();
+        searchMatches = widget.searchMatchesNotifier.value;
+
+        // We only want to rebuild the row if it contains the search matches in
+        // the row have changed.
+        // TODO(kenz): should we check for unordered equality here? Are there
+        // performance repercussions for doing so?
+        if (!collectionEquals(
+            previousSearchMatchesInRow, newSearchMatchesInRow)) setState(() {});
```
[the remark this answers is not in the document]
Comment (kenzieschmoll) [reply packages/devtools_app/lib/src/charts/flame_chart.dart]: done

File: packages/devtools_app/lib/src/charts/flame_chart.dart [hunk 4]
```diff
@@ -477,6 +516,40 @@ class ScrollingFlameChartRowState<V> extends State<ScrollingFlameChartRow>
...
+      addAutoDisposeListener(widget.activeSearchMatchNotifier, () {
+        final containsPreviousActiveSearchMatch =
+            nodeData.contains(activeSearchMatch);
+        final containsNewActiveSearchMatch =
+            nodeData.contains(widget.activeSearchMatchNotifier.value);
+        activeSearchMatch = widget.activeSearchMatchNotifier.value;
+        // We only want to rebuild the row if it contains the previous or new
+        // active search match.
+        if (containsPreviousActiveSearchMatch || containsNewActiveSearchMatch)
+          setState(() {});
```
[the remark this answers is not in the document]
Comment (kenzieschmoll) [reply packages/devtools_app/lib/src/charts/flame_chart.dart:551]: done

File: packages/devtools_app/lib/src/timeline/timeline_controller.dart (hunk 1 above)
Comment (kenzieschmoll) [reply packages/devtools_app/lib/src/timeline/timeline_controller.dart]: done

**Comment by kenzieschmoll** [comment]
Disabled the ESC key listener as discussed offline @terrylucas

**Pull request merged**